### ISSUE TITLE
Rework test suite to use fixtures

### ISSUE THREAD

**Issue by annathepiper** [annathepiper/wp-test-demo-python-selenium#11]
Since I'm attempting to overhaul this test suite to use pytest instead of nose for its test runner, it behooves me to also learn how to use fixtures, a prominent pytest feature.

Reference conversation with @diazona on Mastodon is here:

https://mastodon.murkworks.net/@annathepiper/110731402353125388

Possible action items here to evaluate:

- [ ] BaseTest at this point is pretty much just serving to set my Selenium driver, can I eliminate BaseTest completely by using fixtures to set that up instead? Evaluate using scope=session and autouse=True.
- [ ] Anything else that's a good candidate for a fixture? If I'm chucking all the "verify" type methods into their associated "WP" classes, would it actually gain me anything to turn those into fixtures instead?

**Comment by annathepiper** [comment]
Noting for my reference (and for anybody who happens to come by to check out what I'm doing):

As of last checkin on this branch I was able to implement using fixtures on the main test classes in this suite. But when I tried to do the same steps on the WP-* support classes, each and every one of them bitched that they couldn't find the driver object.

Similar errors were thrown if I tried to use the fixtures to set the ac object for ActionChains, or wp_lib to pull in the few remaining strings I have in the WPTestLib library.

Not sure yet what I'm doing wrong here, since I'm very new to fixtures and not clear yet on how to use them across all classes where I want them. Am I setting the scope wrong? Or autouse?

**Comment by annathepiper** [comment]
Checked in a couple more changes to reflect trying to use the fixtures in WPHomepage.

Here's the error output I get when I try to just run the test_homepage_title test in the test_homepage.py file:

`$ pytest --testdox test_homepage.py::TestHomepage::test_homepage_title
================================================================= test session starts =================================================================
platform win32 -- Python 3.7.2, pytest-7.4.0, pluggy-1.2.0
rootdir: C:\cygwin64\home\AngelaK\workspace\wp-test-demo-python-selenium
plugins: order-1.1.0, testdox-3.0.1
collected 1 item

test_homepage.py                                                                                                                                 [100%]

======================================================================= ERRORS ========================================================================
_________________________________________________ ERROR at setup of TestHomepage.test_homepage_title __________________________________________________

self = <test_homepage.TestHomepage object at 0x0421F3D0>

    def setup_method(self):
        """
        Do setup for the test cases.
        """
        # super().setup_method()
        self.driver.get(self.wp_lib.wp_base_uri)
        # self.wp_homepage = WPHomepage.WPHomepage(self.driver, self.wp_lib)
>       self.wp_homepage = WPHomepage.WPHomepage(self.wp_lib)

test_homepage.py:23:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <WPHomepage.WPHomepage object at 0x0423C5B0>, wp_lib = <WPTestLib.WPTestLib object at 0x0423C610>

    def __init__(self, wp_lib):
        """
        Init method.
        :param driver: Webdriver object
        :param wp_lib: wp_lib object for the helper class to get ids, classes, xpaths
        """
        # self.driver = driver
        self.wp_lib = wp_lib

        # Initialize the menu object
>       self.wp_menu = WPMenu.WPMenu(self.driver)
E       AttributeError: 'WPHomepage' object has no attribute 'driver'

WPHomepage.py:33: AttributeError
=============================================================== short test summary info ===============================================================
ERROR test_homepage.py::TestHomepage::test_homepage_title - AttributeError: 'WPHomepage' object has no attribute 'driver'
================================================================== 1 error in 1.45s ===================================================================
`

**Comment by annathepiper** [comment]
ENLIGHTENMENT DAWNS.

I'm informed that apparently fixtures _can't_ be used on the support classes, fixtures only work on actual test classes and methods.

Reference conversation on Mastodon:

https://mastodon.murkworks.net/@annathepiper/110753873107486900

**Comment by diazona** [comment]
If I may offer a piece of advice: as an educational experience, I'd suggest refactoring your whole test suite so it doesn't use _any_ classes at all - neither test classes nor helper classes. Not to say that classes should never be used in tests, but it seems like so far you've been really attached to structuring your test suite around these classes, which I think is kind of at odds with how pytest is designed, i.e. emphasizing test functions and fixtures, not classes. I'm speculating that if you try removing all classes from these tests, it will force you into a structure that is more typically "pytest-ish" so you can see what that's like. Then, having done that, you might have a better sense of how classes do and don't fit in to the whole thing.

**Comment by annathepiper** [comment]
> 

Yeah, it's built from classes pretty much because I ported over this suite from the Java version, and I'm coming at this with a mindset of having used OOP and classes and things pretty much all of my professional life.

I'm not inherently opposed to trying to see if I can eliminate classes entirely here. But I have a hard time seeing how I can go in that direction without also breaking a lot of code practices I've had drilled into me as an SDET.

Primarily, not having to redeclare the same object or string or whatever in a dozen different places if you need that thing in a dozen different tests. In a small personal project like this it's not so much of a big deal maybe, but in a larger automation framework that would scream "nightmare to maintain" to me.

Still though, a learning experience would be good!

I'll see what I can do.

**Comment by diazona** [comment]
> Yeah, it's built from classes pretty much because I ported over this suite from the Java version, and I'm coming at this with a mindset of having used OOP and classes and things pretty much all of my professional life.

Yeah, that makes sense... honestly, I _have_ been thinking your code looks like Java written with Python syntax :stuck_out_tongue: FWIW I can relate: my first proper programming language was Java (followed by C++, which of course is also somewhat object-oriented), and I went through a phase of doing the same thing when I first started working with Python. And of course one can write perfectly good code without ever leaving "OOP-world", but I think if you really want to learn Python, it's worth breaking away from that so you also have the ability to write - and fluently read - non-OOP code in your toolbox. One of the neat things about Python (which you might have heard) is that it supports a bunch of different programming paradigms: OOP, procedural, functional, async, etc. so if you can become familiar with them you get a lot of flexibility to use the right tool for any job. (Maybe less applicable to testing than general coding, but still, I think it helps not to be limited to one.)

Anyway... there are other ways to reuse values in different tests without inheritance. Like, you could stick them in an importable module: in a typical Python package you'd have some source code for the package itself maybe in an `src/` directory, and you could designate part of that source code as testing helper code (`src/mypackage/testing/constants.py` or whatever) and import that when you need it. Or, there's a package [`pytest-helpers-namespace`](https://pypi.org/project/pytest-helpers-namespace/) which gives you an easy way to make helper code accessible to your tests using pytest itself, without having to structure it as a package. Or you could actually write a version of that yourself; I've done it for a package and it's not too hard. Or, perhaps the easiest way, you could just group the constants under a fixture and have any test that needs it request that fixture.

If there are other SDET habits that you think are pushing you toward keeping classes in the code... well, I don't have SDET training so I'm sort of guessing here, but I do suspect that there will be ways to keep to good habits without using classes. And if not, then sure, maybe the _best_ way to structure these tests does involve some classes - but I do still think that it's worth trying something that might not be a best practice but lets you push the boundaries, which if nothing else might help clarify why the best practices are best practices. And a small personal project like this is the perfect way to do it. IMO you're not learning much if you don't do things "wrong" sometimes. (Maybe this is obvious to you, maybe some of it is wrong... I dunno, this is all just my opinion, hopefully a somewhat informed one, but I figure you'd appreciate at least hearing it.)
